libnavigation-base/src/main/java/com/mapbox/navigation/base/route/NavigationRoute.kt

```diff
@@ -0,0 +1,175 @@
+@file:JvmName("NavigationRouteEx")
+
+package com.mapbox.navigation.base.route
+
+import com.mapbox.api.directions.v5.models.DirectionsResponse
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.DirectionsWaypoint
+import com.mapbox.api.directions.v5.models.LegStep
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.api.directions.v5.models.RouteOptions
+import java.util.UUID
+
+/**
+ * Wraps a route object used across the Navigation SDK features.
+ *
+ * @param directionsResponse the original response that that returned this route object.
+ * @param routeIndex the index of the route that this wrapper tracks
+ * from the collection of routes returned in the original response.
+ * @param routeOptions options used to generate the [directionsResponse]
+ * @param id unique ID of this route entity used for matching it against additional metadata.
+ * Unique ID doesn't mean that the underlying [directionsRoute] is unique.
+ * Multiple [NavigationRoute]s can represent similar or equal [DirectionsRoute].
+ */
+class NavigationRoute(
+    val directionsResponse: DirectionsResponse,
+    val routeIndex: Int,
+    val routeOptions: RouteOptions,
+    val id: String = UUID.randomUUID().toString()
+) {
+
+    /**
+     * [DirectionsRoute] that this [NavigationRoute] represents.
+     */
+    val directionsRoute = directionsResponse.routes()[routeIndex]!!.toBuilder()
+        .requestUuid(directionsResponse.uuid())
+        .routeIndex(routeIndex.toString())
+        .routeOptions(routeOptions)
+        .build()
+
+    /**
+     * Indicates whether some other object is "equal to" this one.
+     */
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as NavigationRoute
+
+        if (directionsRoute != other.directionsRoute) return false
```

Comment:
shouldn't it compare all the fields: `directionsResponse`, `routeIndex`, `routeOptions`, `id` ?

Reply:
I was giving this some thought and I couldn't find a practical application where `NavigationRoute` should not be equal if the wrapped `DirectionsRoute` is equal. It would be inefficient to compare all other routes from the response when they are not tracked by this instance of the `NavigationRoute`.

That said, I can imagine some extreme edge cases like if the equal routes were generated by different endpoints and only after refresh they would return different data.

I think we can extend the equals check to `RouteOptions` and some response properties (but not all `response#routes()) and I'm open to other thoughts on the topic.

Reply:
Do I understand you correctly that it's a performance optimisation? Like we sure that if `directionsRoute` fields are the same then it doesn't make sense to compare the rest of the object?

Reply:
Yes, performance is one aspect of it. The other is that if we have 2 route responses:
- first contains route `A` and route `B`
- second contains route `A` and route `C`

where route `A` is equal between these two requests, and then we have two `NavigationRoute` instances `N1` and `N2` where `N1` track route `A` from the first response and `N2` tracks route `A` from the second response, we don't want the equality to be broken by `B` and `C` which are different across the response objects but are not tracked by the compared instances of the `NavigationRoute`, so we shouldn't consider them for equality.

I added a comment in code to highlight this.